### PR TITLE
Rewrite in Phoenix

### PR DESCRIPTION
**Pull request by KasperVaessen**

### CODE REVIEW

**Review by ThomBreugelmans — COMMENTED** [review]
Running the project with Docker results in the following error:

```
ERROR! Config provider Config.Reader failed with:
** (RuntimeError) environment variable SECRET_KEY_BASE is missing.
You can generate one by calling: mix phx.gen.secret

    /app/releases/0.1.0/runtime.exs:31: (file)
    (elixir 1.17.3) src/elixir.erl:386: :elixir.eval_external_handler/3
    (stdlib 6.1.2) erl_eval.erl:904: :erl_eval.do_apply/7
    (stdlib 6.1.2) erl_eval.erl:648: :erl_eval.expr/6
    (stdlib 6.1.2) erl_eval.erl:271: :erl_eval.exprs/6
    (elixir 1.17.3) src/elixir.erl:364: :elixir.eval_forms/4
    (elixir 1.17.3) lib/module/parallel_checker.ex:112: Module.ParallelChecker.verify/1
    (elixir 1.17.3) lib/code.ex:572: Code.validated_eval_string/3

Runtime terminating during boot ({#{message=><<101,110,118,105,114,111,110,109,101,110,116,32,118,97,114,105,97,98,108,101,32,83,69,67,82,69,84,95,75,69,89,95,66,65,83,69,32,105,115,32,109,105,115,115,105,110,103,46,10,89,111,117,32,99,97,110,32,103,101,110,101,114,97,116,101,32,111,110,101,32,98,121,32,99,97,108,108,105,110,103,58,32,109,105,120,32,112,104,120,46,103,101,110,46,115,101,99,114,101,116,10>>,'__struct__'=>'Elixir.RuntimeError','__exception__'=>true},[{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.1.0/runtime.exs"},{line,31}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,386},{error_info,#{module=>'Elixir.Exception'}}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,904}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,648}]},{erl_eval,exprs,6,[{file,"erl_eval.erl"},{line,271}]},{elixir,eval_forms,4,[{file,"src/elixir.erl"},{line,364}]},{'Elixir.Module.ParallelChecker',verify,1,[{file,"lib/module/parallel_checker.ex"},{line,112}]},{'Elixir.Code',validated_eval_string,3,[{

Crash dump is being written to: erl_crash.dump...done
```

I have supplied the (afaik) required environment variables `session` and `leaderboard_id`.

**Comment by KasperVaessen** [comment]
> Running the project with Docker results in the following error:
> 
> ```
> ERROR! Config provider Config.Reader failed with:
> ** (RuntimeError) environment variable SECRET_KEY_BASE is missing.
> You can generate one by calling: mix phx.gen.secret
> 
>     /app/releases/0.1.0/runtime.exs:31: (file)
>     (elixir 1.17.3) src/elixir.erl:386: :elixir.eval_external_handler/3
>     (stdlib 6.1.2) erl_eval.erl:904: :erl_eval.do_apply/7
>     (stdlib 6.1.2) erl_eval.erl:648: :erl_eval.expr/6
>     (stdlib 6.1.2) erl_eval.erl:271: :erl_eval.exprs/6
>     (elixir 1.17.3) src/elixir.erl:364: :elixir.eval_forms/4
>     (elixir 1.17.3) lib/module/parallel_checker.ex:112: Module.ParallelChecker.verify/1
>     (elixir 1.17.3) lib/code.ex:572: Code.validated_eval_string/3
> 
> Runtime terminating during boot ({#{message=><<101,110,118,105,114,111,110,109,101,110,116,32,118,97,114,105,97,98,108,101,32,83,69,67,82,69,84,95,75,69,89,95,66,65,83,69,32,105,115,32,109,105,115,115,105,110,103,46,10,89,111,117,32,99,97,110,32,103,101,110,101,114,97,116,101,32,111,110,101,32,98,121,32,99,97,108,108,105,110,103,58,32,109,105,120,32,112,104,120,46,103,101,110,46,115,101,99,114,101,116,10>>,'__struct__'=>'Elixir.RuntimeError','__exception__'=>true},[{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.1.0/runtime.exs"},{line,31}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,386},{error_info,#{module=>'Elixir.Exception'}}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,904}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,648}]},{erl_eval,exprs,6,[{file,"erl_eval.erl"},{line,271}]},{elixir,eval_forms,4,[{file,"src/elixir.erl"},{line,364}]},{'Elixir.Module.ParallelChecker',verify,1,[{file,"lib/module/parallel_checker.ex"},{line,112}]},{'Elixir.Code',validated_eval_string,3,[{
> 
> Crash dump is being written to: erl_crash.dump...done
> ```
> 
> I have supplied the (afaik) required environment variables `session` and `leaderboard_id`.

Ah yes indeed you also need to supply a SECRET_KEY_BASE env variable. I sent you one in private which you can use.

**Comment by ThomBreugelmans** [comment]
I see, with that `SECRET_KEY_BASE` it works, what is the use for this environment variable? (would be nice to have this reflected in the README as well so it is known how to run it)
Little side note, if this is required this will need to be added to the cluster config when it is pushed to main so that our cloud instance will be able to run.

File: lib/aoch.ex [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        Map.put(map, :star1, AoCHWeb.format_time(map[:first_star]))
+      else
+        map
+      end
+
+    if map[:second_star] do
+      Map.put(map, :star2, "+ " <> AoCHWeb.format_time(map[:second_star]))
+    else
+      map
+    end
```
Comment (ThomBreugelmans) [inline lib/aoch.ex:80]: Very much nagging, however would it be possible for this section to `null` the `star2` field if the first is submitted but the second is not yet (like with the `second_star` field)?

example:
image

*possible suggestion, however never used elixir/phoenix so please do not blindly accept*
```suggestion
    map =
      if map[:first_star] do
        Map.put(map, :star1, AoCHWeb.format_time(map[:first_star]))
      else
        map
      end

    if map[:second_star] do
      Map.put(map, :star2, "+ " <> AoCHWeb.format_time(map[:second_star]))
    else
      Map.put(map, :star2, nil)
    end
```

**Review by ThomBreugelmans — CHANGES_REQUESTED** [review]

File: lib/aoch.ex (hunk 1 above)
Comment (KasperVaessen) [reply lib/aoch.ex:80]: Yes that ofcourse can be done. However, I did it like this on purpose to stay consistent with the old api.

**Comment by KasperVaessen** [comment]
> I see, with that `SECRET_KEY_BASE` it works, what is the use for this environment variable? (would be nice to have this reflected in the README as well so it is known how to run it)
> 
> Little side note, if this is required this will need to be added to the cluster config when it is pushed to main so that our cloud instance will be able to run.

This key is used by the Phoenix framework as base for security purposes. Our app has no user interaction, so it might not be 100% neccesary, however it is considered best practise to always include it. I will add it to the readme, just like the other two env variables

File: lib/aoch.ex (hunk 1 above)
Comment (ThomBreugelmans) [reply lib/aoch.ex:80]: That is true, i'll leave it as is for now and if someone really wants/would like that to change they can create a PR/issue for it.

**Review by ThomBreugelmans — APPROVED** [review]

**Pull request merged**